File: backend/pkg/providers/gemini.go
package providers

import (
	"bufio"
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"strings"
	"time"
)

type GeminiClient struct {
	APIKey string
	Model  string
	client *http.Client
}

func NewGeminiClient(apiKey, model string) *GeminiClient {
	if model == "" {
		model = "gemini-2.5-flash"
	}
	return &GeminiClient{
		APIKey: apiKey,
		Model:  model,
		client: &http.Client{
			Timeout: 10 * time.Minute,
		},
	}
}

type GeminiBlob struct {
	MimeType string `json:"mimeType"`
	Data     string `json:"data"`
}

type GeminiPart struct {
	Text         string        `json:"text,omitempty"`
	Thought      bool          `json:"thought,omitempty"`
	InlineData   *GeminiBlob   `json:"inlineData,omitempty"`
	FunctionCall *FunctionCall `json:"functionCall,omitempty"`
	FunctionResp *FunctionResp `json:"functionResponse,omitempty"`
}

type FunctionCall struct {
	ID   string         `json:"-"`
	Name string         `json:"name"`
	Args map[string]any `json:"args"`
}

type FunctionResp struct {
	ID       string         `json:"-"`
	Name     string         `json:"name"`
	Response map[string]any `json:"response"`
}

type GeminiContent struct {
	Role  string       `json:"role"`
	Parts []GeminiPart `json:"parts"`
}

type GeminiRequest struct {
	Contents          []GeminiContent  `json:"contents"`
	SystemInstruction *GeminiContent   `json:"systemInstruction,omitempty"`
	Tools             []map[string]any `json:"tools,omitempty"`
}

type GeminiCandidate struct {
	Content struct {
		Parts []GeminiPart `json:"parts"`
		Role  string       `json:"role"`
	} `json:"content"`
	FinishReason string `json:"finishReason"`
}

type GeminiResponse struct {
	Candidates []GeminiCandidate `json:"candidates"`
	Error      *struct {
		Code    int    `json:"code"`
		Message string `json:"message"`
		Status  string `json:"status"`
	} `json:"error,omitempty"`
}

type StreamCallbacks struct {
	OnTextChunk    func(string)
	OnThoughtChunk func(string)
	OnFunctionCall func(FunctionCall)
}

func (g *GeminiClient) StreamGenerateContent(ctx context.Context, req GeminiRequest, cb StreamCallbacks) error {
	url := fmt.Sprintf("https://generativelanguage.googleapis.com/v1beta/models/%s:streamGenerateContent?alt=sse&key=%s", g.Model, g.APIKey)

	reqBody, err := json.Marshal(req)
	if err != nil {
		return fmt.Errorf("failed to marshal gemini request: %w", err)
	}

	httpReq, err := http.NewRequestWithContext(ctx, "POST", url, bytes.NewReader(reqBody))
	if err != nil {
		return fmt.Errorf("failed to create http request: %w", err)
	}
	httpReq.Header.Set("Content-Type", "application/json")

	resp, err := g.client.Do(httpReq)
	if err != nil {
		return fmt.Errorf("gemini request failed: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		body, _ := io.ReadAll(resp.Body)
		return fmt.Errorf("gemini api error (HTTP %d): %s", resp.StatusCode, string(body))
	}

	reader := bufio.NewReader(resp.Body)
	for {
		select {
		case <-ctx.Done():
			return ctx.Err()
		default:
		}

		line, err := reader.ReadString('\n')
		if err != nil {
			if err == io.EOF {
				break
			}
			return err
		}

		line = strings.TrimSpace(line)
		if strings.HasPrefix(line, "data: ") {
			jsonStr := strings.TrimPrefix(line, "data: ")
			var geminiResp GeminiResponse
			if err := json.Unmarshal([]byte(jsonStr), &geminiResp); err == nil {
				if geminiResp.Error != nil {
					return fmt.Errorf("gemini streaming error: %s (code %d)", geminiResp.Error.Message, geminiResp.Error.Code)
				}
				for _, cand := range geminiResp.Candidates {
					for _, part := range cand.Content.Parts {
						if part.Thought && part.Text != "" {
							if cb.OnThoughtChunk != nil {
								cb.OnThoughtChunk(part.Text)
							}
						} else if part.Text != "" {
							if cb.OnTextChunk != nil {
								cb.OnTextChunk(part.Text)
							}
						}
						if part.FunctionCall != nil {
							if cb.OnFunctionCall != nil {
								cb.OnFunctionCall(*part.FunctionCall)
							}
						}
					}
				}
			}
		}
	}

	return nil
}

File: providers/gemini.js
// Google Gemini Provider Extension

function cleanToolResponse(resp) {
  if (!resp || typeof resp !== "object") return resp;
  try {
    const clone = JSON.parse(JSON.stringify(resp));
    const clean = (obj) => {
      if (!obj || typeof obj !== "object") return;
      for (const k of Object.keys(obj)) {
        if (k === "inlineImage" || k === "inlineData") {
          obj[k] = "[image binary data]";
        } else if (k === "data" && typeof obj[k] === "string" && obj[k].length > 200) {
          obj[k] = "[binary data]";
        } else if (typeof obj[k] === "object") {
          clean(obj[k]);
        }
      }
    };
    clean(clone);
    return clone;
  } catch (_) {
    return resp;
  }
}

export default {
  id: "gemini",
  name: "Google Gemini",
  description: "Official Google Gemini API supporting streaming, reasoning, and tool configurations.",
  
  requirements: [
    { key: "apiKey", label: "Gemini API Key", type: "password", required: true, placeholder: "AIzaSy..." },
    { key: "model", label: "Model Identifier", type: "text", required: true, default: "gemini-2.5-flash", placeholder: "gemini-2.5-flash, gemini-2.0-flash, gemini-2.5-pro" },
    { key: "temperature", label: "Temperature", type: "number", required: false, default: 0.7, min: 0, max: 2, step: 0.1 }
  ],

  async *stream({ apiKey, model, temperature, messages, tools, systemInstruction, abortSignal }) {
    const targetModel = model || "gemini-2.5-flash";
    const url = `https://generativelanguage.googleapis.com/v1beta/models/${targetModel}:streamGenerateContent?alt=sse&key=${apiKey}`;

    // Convert messages to Gemini format
    const contents = (messages || []).map(m => {
      let parts = [];
      if (Array.isArray(m.parts)) {
        parts = m.parts.map(p => {
          if (p.functionCall) {
            return {
              functionCall: {
                name: p.functionCall.name,
                args: p.functionCall.args || {}
              }
            };
          } else if (p.functionResponse) {
            const rawResp = typeof p.functionResponse.response === 'object' && p.functionResponse.response !== null
              ? p.functionResponse.response
              : { result: p.functionResponse.response };
            return {
              functionResponse: {
                name: p.functionResponse.name,
                response: cleanToolResponse(rawResp)
              }
            };
          } else if (p.inlineData) {
            return {
              inlineData: {
                mimeType: p.inlineData.mimeType,
                data: p.inlineData.data
              }
            };
          } else if (p.text) {
            return { text: p.text };
          }
          return { text: typeof p === 'string' ? p : JSON.stringify(p) };
        });
      } else if (typeof m.content === "string") {
        parts = [{ text: m.content }];
      } else if (Array.isArray(m.content)) {
        parts = m.content.map(c => typeof c === "string" ? { text: c } : { text: c.text || JSON.stringify(c) });
      } else {
        parts = [{ text: JSON.stringify(m.content || "") }];
      }

      return {
        role: m.role === "assistant" || m.role === "model" ? "model" : "user",
        parts
      };
    });

    const requestBody = {
      contents,
      generationConfig: {
        temperature: temperature !== undefined ? Number(temperature) : 0.7
      }
    };

    if (systemInstruction) {
      requestBody.systemInstruction = {
        parts: [{ text: typeof systemInstruction === 'string' ? systemInstruction : JSON.stringify(systemInstruction) }]
      };
    }

    if (tools && tools.length > 0) {
      requestBody.tools = [{
        functionDeclarations: tools.map(t => ({
          name: t.name,
          description: t.description,
          parameters: t.parameters || t.inputSchema || { type: "object", properties: {} }
        }))
      }];
    }

    const response = await fetch(url, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(requestBody),
      signal: abortSignal
    });

    if (!response.ok) {
      const errText = await response.text();
      throw new Error(`Gemini API Error (${response.status}): ${errText}`);
    }

    const reader = response.body.getReader();
    const decoder = new TextDecoder();
    let buffer = "";

    while (true) {
      const { done, value } = await reader.read();
      if (done) break;
      buffer += decoder.decode(value, { stream: true });
      const lines = buffer.split("\n");
      buffer = lines.pop() || "";

      for (const line of lines) {
        const trimmed = line.trim();
        if (!trimmed || !trimmed.startsWith("data: ")) continue;
        try {
          const data = JSON.parse(trimmed.slice(6));
          const candidate = data.candidates?.[0];
          const parts = candidate?.content?.parts || [];
          for (const part of parts) {
            if (part.thought) {
              yield { type: "thought", text: part.text || part.thought };
            } else if (part.functionCall) {
              yield {
                type: "functionCall",
                name: part.functionCall.name,
                args: part.functionCall.args,
                callId: part.functionCall.id || ("call_" + Math.random().toString(36).substring(2, 8))
              };
            } else if (part.text) {
              yield { type: "text", text: part.text };
            }
          }
          if (candidate?.finishReason) {
            try { reader.cancel(); } catch(e) {}
            return;
          }
        } catch (e) {}
      }
    }
  }
};
